third-party/rsutils/src/network-adapter-watcher.cpp

```diff
@@ -0,0 +1,234 @@
+
+                  int const timeout = -1;  // negative value = infinite
+                  int const nfds = 2;
+                  pollfd fds[nfds];
+                  memset( fds, 0, sizeof( fds ) );
+                  fds[0].fd = _done;
+                  fds[0].events = POLLIN;
+                  fds[1].fd = _socket;
+                  fds[1].events = POLLIN;
+
+                  while( true )
+                  {
+                      fds[0].revents = 0;
+                      fds[1].revents = 0;
+                      auto rv = poll( fds, nfds, timeout );
+                      if( rv <= 0 )
+                      {
+                          LOG_ERROR( "poll failed: " << rv );
+                          break;
+                      }
+
+                      if( fds[0].revents != 0 )
+                          break;  // we're done
+
+                      LOG_DEBUG( "network adapter changes detected!" );
+
+                      char buf[4096];
+                      struct sockaddr_nl snl;
+                      iovec iov = { buf, sizeof buf };
+                      msghdr msg = { (void *)&snl, sizeof snl, &iov, 1, NULL, 0, 0 };
+                      int status = recvmsg( _socket, &msg, 0 );
+                      if( status < 0 )
+                      {
+                          // Socket non-blocking so bail out once we have read everything
+                          if( errno == EWOULDBLOCK || errno == EAGAIN )
+                              continue;
+
+                          // Anything else is an error
+                          LOG_DEBUG( "read_netlink: Error recvmsg: " << status );
+                          break;
+                      }
+                      if( status == 0 )
+                          LOG_DEBUG( "read_netlink: EOF" );
+
+                      // We need to handle more than one message per recvmsg
+                      for( auto h = (struct nlmsghdr *)buf; NLMSG_OK( h, (unsigned int)status );
+                           h = NLMSG_NEXT( h, status ) )
+                      {
+                          // Finish reading
+                          if( h->nlmsg_type == NLMSG_DONE )
+                              continue;
+
+                          // Message is some kind of error
+                          if( h->nlmsg_type == NLMSG_ERROR )
+                          {
+                              LOG_DEBUG( "read_netlink: got NLMSG_ERROR" );
+                              continue;
+                          }
+                      }
+
+                      // Call our clients once for all the messages... outside the loop
+                      callbacks.raise();
+                  }
+
+                  LOG_DEBUG( "exiting network adapter watcher" );
+              } )
+#endif
+    {
+        LOG_DEBUG( "network adapter watcher singleton is up" );
+    }
+
+    ~network_adapter_watcher_singleton()
+    {
+#ifdef _WIN32
+        if( _th.joinable() )
+        {
+            SetEvent( _done );
```

Comment:
Don't we need to set `_done` for detached threads as well?

Reply:
Detached threads?
If it's not joinable then it's not running, and no reason I can think of to set this event... right?

Reply:
Yes, you have used detached threads elsewhere so I mixed the two...